File: 리트코드/0241-different-ways-to-add-parentheses/0241-different-ways-to-add-parentheses.java
class Solution {
    public List<Integer> diffWaysToCompute(String expression) {
        List<Integer> result = getCalculateResult(0, expression.length() - 1, expression);

        return result;
    }

    public List<Integer> getCalculateResult(int left, int right, String expression) {
        List<Integer> result = new ArrayList<>();
        String substring = expression.substring(left, right + 1);

        if (!substring.contains("*") && !substring.contains("-") && !substring.contains("+")) {
            result.add(Integer.parseInt(substring));
            return result;
        }

        for (int i = left; i <= right; i++) {
            if (!Character.isDigit(expression.charAt(i))) {
                List<Integer> leftList = getCalculateResult(left, i - 1, expression);
                List<Integer> rightList = getCalculateResult(i + 1, right, expression);

                for (Integer l : leftList) {
                    for (Integer r : rightList) {
                        if (expression.charAt(i) == '+') {
                            result.add(l + r);
                        } else if (expression.charAt(i) == '-') {
                            result.add(l - r);
                        } else if (expression.charAt(i) == '*') {
                            result.add(l * r);
                        }
                    }
                }
            }
        }

        return result;
    }
}

File: 리트코드/0241-different-ways-to-add-parentheses/0241-different-ways-to-add-parentheses.kt
class Solution {
    val memo: MutableMap<String,List<Int>> = mutableMapOf()
    fun diffWaysToCompute(expression: String): List<Int> {
        if (memo.containsKey(expression)) {
            return memo[expression]!!
        }

        val result: MutableList<Int> = mutableListOf()
        for(i in expression.indices){

            val c = expression[i]
            if (c in "+-*") {

                val left = diffWaysToCompute(expression.substring(0, i))
                val right = diffWaysToCompute(expression.substring(i + 1))

                for (l in left) {
                    for (r in right){
                        when (c) {
                            '+' -> result.add(l + r)
                            '-' -> result.add(l - r)
                            '*' -> result.add(l * r)
                        }
                    }
                }
            }
        }
        
        if (result.isEmpty()) {
            result.add(expression.toInt())
        }
        
        memo[expression] = result
        return result
    }
}